### PR TITLE
Cpu profiler

### PR DESCRIPTION
**Pull request by ryzhyk**
Cleanup in preparation for Z-set refactoring; initial CPU profiler implementation.

### CODE REVIEW

File: src/algebra/zset/tests.rs [hunk 1]
```diff
@@ -254,27 +254,9 @@ fn zset_tuple_tests() {
...
-    assert_eq!(
-        finite_map! {
-            (-1, 0) => CheckedI64::new(9),
-            (0, 0) => CheckedI64::new(9),
-        },
-        j,
-    );
 }
 
+/*
```
Comment (mbudiu-vmw) [inline src/algebra/zset/tests.rs:259]: why not delete it altogether?

File: src/profile/cpu.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+#[derive(Clone, Default)]
+pub struct CPUProfiler(Rc<RefCell<CPUProfilerInner>>);
+
+impl CPUProfiler {
+    /// Create a new CPU profiler instance.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Attach CPU profiler to a circuit.  The profiler will start circuit's CPU
```
Comment (mbudiu-vmw) [inline src/profile/cpu.rs:29]: start -> measure?

File: src/profile/cpu.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    pub fn operator_profile(&self, node: &GlobalNodeId) -> Option<OperatorCPUProfile> {
+        if let Ok(this) = self.0.try_borrow() {
+            this.operators.get(node).cloned()
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Default)]
```
Comment (mbudiu-vmw) [inline src/profile/cpu.rs:53]: I think this should be before the impl for readability.

File: src/profile/cpu.rs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                    op_profile.total_time += duration;
+                };
+            }
+            _ => (),
+        }
+    }
+}
+
+/// Per-operator CPU profile.
+#[derive(Clone, Default)]
```
Comment (mbudiu-vmw) [inline src/profile/cpu.rs:83]: For code readability I prefer the definition before the first use.

File: src/profile/cpu.rs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            SchedulerEvent::EvalEnd { node } => {
+                if let Some(start_time) = self.start_times.remove(node.global_id()) {
+                    let duration = Instant::now().duration_since(start_time);
+                    let op_profile = self
+                        .operators
+                        .entry(node.global_id().clone())
+                        .or_insert_with(Default::default);
+                    op_profile.invocations += 1;
+                    op_profile.total_time += duration;
+                };
```
Comment (mbudiu-vmw) [inline src/profile/cpu.rs:75]: So the start time can be missing if you toggle profiling while running?

Comment (mbudiu-vmw) [reply src/profile/cpu.rs:75]: But in that case the results may be wrong if you start/stop profiling - you will use perhaps an old start time.

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: src/profile/cpu.rs (hunk 5 above)
Comment (ryzhyk) [reply src/profile/cpu.rs:75]: If we add the ability to stop and re-start profiling, start times will need to be cleared for this reason.

**Pull request merged**